llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -3155,56 +3165,85 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+static void
+restoreStack(BinaryStreamReader &Reader, StringRef &Data,
+             std::stack<std::pair<StringRef, unsigned>> &StackOfNodes) {
+  auto DataAndOffset = StackOfNodes.top();
+  Reader = BinaryStreamReader(DataAndOffset.first, llvm::endianness::little);
+  Data = DataAndOffset.first;
+  Reader.setOffset(DataAndOffset.second);
+  StackOfNodes.pop();
+}
 
-  if (Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag())
-    StartTagCallback(*MaybeTag, AbbrevIdx);
-  else
-    return MaybeTag.takeError();
+Error DIEVisitor::visitDIERef(unsigned AbbrevIdx,
+                              ArrayRef<DIEDataRef> &DIEChildrenStack) {
 
-  Expected<bool> MaybeHasChildren = AbbrevReader.readHasChildren();
-  if (!MaybeHasChildren)
-    return MaybeHasChildren.takeError();
+  std::stack<std::pair<StringRef, unsigned>> StackOfNodes;
+  auto Data = DIEChildrenStack.empty() ? StringRef()
+                                       : DIEChildrenStack.front().getData();
+  BinaryStreamReader Reader(Data, llvm::endianness::little);
+  while (!DistinctReader.empty()) {
 
-  if (auto E = visitDIEAttrs(AbbrevReader, DataReader, DIEData))
-    return E;
+    AbbrevEntryReader AbbrevReader =
+        getAbbrevEntryReader(AbbrevEntries, AbbrevIdx);
 
-  if (!*MaybeHasChildren) {
-    EndTagCallback(false /*HadChildren*/);
-    return Error::success();
-  }
+    if (Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag())
+      StartTagCallback(*MaybeTag, AbbrevIdx);
+    else
+      return MaybeTag.takeError();
 
-  while (true) {
-    Expected<uint64_t> ChildAbbrevIdx = readAbbrevIdx(DistinctReader);
-    if (!ChildAbbrevIdx)
-      return ChildAbbrevIdx.takeError();
+    Expected<bool> MaybeHasChildren = AbbrevReader.readHasChildren();
+    if (!MaybeHasChildren)
+      return MaybeHasChildren.takeError();
 
-    if (*ChildAbbrevIdx == getEndOfDIESiblingsMarker())
-      break;
+    if (auto E = visitDIEAttrs(AbbrevReader, Reader, Data))
+      return E;
 
-    if (*ChildAbbrevIdx == getDIEInAnotherBlockMarker()) {
-      if (auto E = visitDIERef(DIEChildrenStack.front()))
-        return E;
-      DIEChildrenStack = DIEChildrenStack.drop_front();
-      continue;
+    if (!*MaybeHasChildren) {
+      if (!StackOfNodes.empty() && Reader.empty())
+        restoreStack(Reader, Data, StackOfNodes);
+      EndTagCallback(false /*HadChildren*/);
     }
 
-    if (auto E =
-            visitDIERef(DataReader, *ChildAbbrevIdx, DIEData, DIEChildrenStack))
-      return E;
-  }
+    while (true) {
+      auto NextAbbrevIdx = readAbbrevIdx(DistinctReader);
 
-  EndTagCallback(true /*HadChildren*/);
+      if (!NextAbbrevIdx)
+        return NextAbbrevIdx.takeError();
+
+      if (*NextAbbrevIdx == getEndOfDIESiblingsMarker()) {
+        EndTagCallback(true /*HadChildren*/);
+        if (!StackOfNodes.empty() && Reader.empty())
+          restoreStack(Reader, Data, StackOfNodes);
+
+        if (DistinctReader.empty())
+          break;
```

Comment:
instead of `if (empty) break; continue` why not make the loop be `if (!DistinctReader.empty())`?